### ISSUE TITLE
LIFO connection pool does not give every connection equal chance of being chosen

### ISSUE THREAD

**Issue by ktaekwon000** [go-redis/redis#1819]
## Introduction

Currently, the connection pool of go-redis is implemented as a LIFO stack. This has caused some problem within our organisation when using our library, where we use a single redis client to connect to multiple proxy servers for concurrency purposes. (tldr: each connection in the go-redis connection pool is a connection to a single redis proxy, which then acts as an abstraction for a redis cluster behind it. One client connects to multiple redis proxies and ideally the load from the client will be balanced between the proxies). 

From some testing, we found that with the LIFO connection pool, the connections at the end of the array have a disproportionate chance of being selected rather than other connections in the connection pool. This causes an issue with our setup, where we depend on the connection pool for load balancing.

## Expected Behavior

Each connection in the connection pool has similar chances of being used.

## Current Behavior

Some connections in the connection pool are used less often than others.

## Possible Solution

We have tested both of these solutions in our environment and verified that both of these solutions fix the issue at hand.

### Option 1: Randomized, simpler, involves minimal changes

![image](https://user-images.githubusercontent.com/65844703/125382695-b97eee80-e3c8-11eb-954d-f73b7b9ae41b.png)
When a connection is returned to the connection pool, put it in an arbitary slot in the connection pool. The next connection to be returned is randomly chosen from the connection pool too, with this logic.

The primary downside of this logic is that the current tests will not apply to this kind of connection pool any more, as the existing tests depend on the ordering of the connections in the pool being stable.

### Option 2: Convert idleConns into a linked list, convert the connection pool into a FIFO queue

![image](https://user-images.githubusercontent.com/65844703/125383126-8426d080-e3c9-11eb-9071-2a89a57348e7.png)
![image](https://user-images.githubusercontent.com/65844703/125383198-a1f43580-e3c9-11eb-8715-d97f3a66acb0.png)
When retrieving an idle connection from the connection pool, we can retrieve the connection in a FIFO manner, so that the connection pool is implemented in a round robin fashion, where for a connection that has been put into a connection pool to be get again, all existing connections in the pool must be used before that.

The primary downside to this is the complicated nature of the code change, as changing the data structure of idleConns will cause extensive changes in the connection pool. However, as the order of the connections in the connection pool is deterministic, the existing tests can be modified for the new type of connection pool.

## Steps to Reproduce

1. Have a DNS resolver that resolves multiple IP addresses from a single hostname
2. Create a redis client with poolSize set in the options.
3. Use the connection pool with some commands, like Get or Set. Note that the commands are not distributed among the connections in the connection pool, especially in sequential queries.

## Context (Environment)

In our organisation, each connection in the go-redis connection pool is a connection to a single redis proxy, which then acts as an abstraction for a redis cluster behind it. One client connects to multiple redis proxies and ideally the load from the client will be balanced between the proxies. This issue is affecting our use of go-redis to balance the load between the proxies.

## Detailed Description

Due to the LIFO nature of the pool, each connection in the connection pool has a difference chance of being chosen, and some connections are chosen more often than others. In the load balancing scenario mentioned above, this causes a problem where load balancing does not work.

## Possible Implementation

I can submit a PR for solution 1 mentioned above, if other devs want it.

**Comment by vmihailenco** [comment]
Thanks for the detailed issue - it is very well written :+1: 

The LIFO connection pool was chosen so some connections are completely unused and can be closed & removed from the pool when traffic is low. LIFO is also slightly simpler/faster to implement than FIFO.

>Option 2: Convert idleConns into a linked list, convert the connection pool into a FIFO queue

We could continue to use a slice and `copy` (database/sql does this so apparently it is fast enough):

```go
conn := db.freeConn[0] // get the first free conn
copy(db.freeConn, db.freeConn[1:]) // adjust the slice
```

So I would prefer we do something similar.

@monkey92t what do you think?

**Comment by monkey92t** [comment]
Yes, only a small amount of code needs to be modified.

```go
func (p *ConnPool) popIdle() *Conn {
	if len(p.idleConns) == 0 {
		return nil
	}
	cn := p.idleConns[0]
	copy(p.idleConns, p.idleConns[1:])
	p.idleConns = p.idleConns[:len(p.idleConns)-1]
	p.idleConnsLen--
	p.checkMinIdleConns()
	return cn
}
```

**Comment by vmihailenco** [comment]
@ktaekwon000 please proceed if you are fine with the suggested solution. I think the LIFO order should remain the default, but it should be possible to switch to FIFO order somehow...

**Comment by ktaekwon000** [comment]
Thanks, the suggested solution looks good. I'm currently thinking of allowing the user to switch via a parameter passed into redis.Options and then pool.Options.

**Comment by monkey92t** [comment]
fixed by #1820
